### DISCUSSION TITLE
Settings Screen Doesn't Allow/Provide Return to Main Screen

### DISCUSSION THREAD

**Discussion by daveclifford** [DCC-EX/EX-T3-WiFi#86] [Q&A]
I finished assembling the T3 Throttle and went through a test functions/throttle famiiarization phase.

When I selected the Settings Screen (Loco UI/Screen Rotation), I can't go back to the Main Screen, unless I turn the Throttle off snd then back on. 

Is this the correct behavior or did I miss something ?

Dave

**Comment by alex-code** [comment]
Hi Dave,

Touching the menu icon on the top right will always take you back to the menu.
With settings this is what also triggers the save to remember the settings so if you powered off any settings you selected would change back to the previous ones.

**Reply by daveclifford** [reply]
Alex,

I touch the 3 bars in the upper right corner of the touch screen and nothing happens. Maybe the touch screen is defective.  All other buutons work.

When powered off, and then powered on, the screen shown  has 4 sections (Loco, Accessories, Power and Settings).

Under Settings, I select the Settings button.

The screen changes to a screen of Loco UI Options,  Touching the three bar icon doesn't do anything.

Dave

**Reply by alex-code** [reply]
hmm, if you go to `By Address`, `By Name`, `By Group` or `Program` and press the menu icon does it go back?

**Reply by daveclifford** [reply]
NO - there is no loco defined.

I have not attached to the DCC CS yet.

Could this be the reason that the T3 Throttle doesn't  go through its hoops?

**Reply by alex-code** [reply]
Pressing the menu icon should always take you back to the menu from any other display.

If you go to `Settings` and change the screen rotation does the menu icon work then?

If not we might need to add a `Serial.println()` to see if the touch is being is being detected.

**Reply by daveclifford** [reply]
(as an aside - it is 0138 AM Sunday 2 April here in Aurora, CO - thw Rocky Mountains  18 miles to our west are sleeping)

I went to Settings and changed the screen rotation - the menu icon worked (in lower left of screen) and the image returned to the main  image - inverted of course.

Could the ribbon cable ( possibly being not seated properly) cause this behavior? I didn't think so, since the icons were fine, positioning was fine,etc.

Maybe the touch display is too tight in the top cover or defective.

Otherwise, the throttle is excellent so far - thoughtful logic, icons and function. Great job!

Dave

**Reply by daveclifford** [reply]
A 3.5" screen  in USD is $39.90 USD and $21 shipping via DHL 

Dave

**Reply by daveclifford** [reply]
I found a US distributor  - Mouser and shipping is only $7.99

**Reply by daveclifford** [reply]
I hate to have to do this, but it looks like I will have to disassemble the throttle to determine the cause of the touch screen failure. 

It might be a case that the screen opening is  too tight and impinged on the screen edge. Only a guess!

I am making another throttle, so I have a  new un-used touchscreen  to try out.

Stay tuned

Dave

**Reply by alex-code** [reply]
You could try adding a `Serial.println("touch");` on line https://github.com/DCC-EX/EX-T3-WiFi/blob/master/src/main.cpp#L442 to see if a touch is being detected.

**Reply by daveclifford** [reply]
Alex,

Took your advice and added  **Serial.println("touch");** to main.cpp, then compiled, uploaded and then monitored.  

When I touch the upper right icon on the Screen Rotation screen, the monitor shows "touch
177168". 

**Each touch** on the screen anywhere displays a monitor "touch nnnnnn" character string.

Also, when I turn the throttle off/then on, the screen inverts.

NOTE:  I am NOT connected to the DCC CS.

Dave

**Reply by alex-code** [reply]
The number is the free ram available, I used it to help debug mem leaks.

Settings won't be saved unless you can exit back to the main menu (without power cycling).

On line https://github.com/DCC-EX/EX-T3-WiFi/blob/master/src/main.cpp#L445 you could try adding
```cpp
Serial.print("X: ");
Serial.print(points[0].x);
Serial.print(" - Y: ");
Serial.println(points[0].y);
```
So we can see what coordinates it's detecting.

**Comment by daveclifford** [comment]
lines 443-449 now are:

if (touches) {
     // Remember the original touch points so we can use later calls to check for swipes
     // Serial.println("touch");
     Serial.print("X: ");
     Serial.print(points[0].x);
     Serial.print(" - Y: ");
     Serial.println(points[0].y);

-------------PIO log below ---------------------------------------------------

Building in release mode
Compiling .pio/build/firebeetle2/src/main.cpp.o
src/main.cpp:23:21: warning: extra tokens at end of #include directive
 #include <WiFiUI.h> Serial.println("touch");
                     ^~~~~~
src/main.cpp: In function 'void loop()':
**src/main.cpp:447:19: error: 'points' was not declared in this scope**
      Serial.print(points[0].x);
                   ^~~~~~
src/main.cpp:447:19: note: suggested alternative: 'printf'
      Serial.print(points[0].x);
                   ^~~~~~
                   printf
*** [.pio/build/firebeetle2/src/main.cpp.o] Error 1
======================= [FAILED] Took 7.29 seconds =

Alex,
I don't know your code well enough to fix the error.

Can I assume the touchscreen hw is working due to the free ram value returned when pushing the return to main screen icon in upper left ?  - or is it too soon to tell???

Dave

**Reply by daveclifford** [reply]
ignore  "src/main.cpp:23:21: warning: extra tokens at end of #include directive"

In editing,  I pasted a string at the end of an include -removed! 

my bad!!!!!

Dave

**Reply by alex-code** [reply]
The `Serial` debug bit needs to be after the `memcpy`.
It seems a touch is being detected but it'll help to see the x & y.

**Reply by daveclifford** [reply]
After success in compiling the  code  (in the correct spot), 

At the Screen Rotation (2/2) and pushing the return icon (upper right), the following character stream occurred:

X: 282 - Y: 18
177176
X: 282 - Y: 18
177176
X: 282 - Y: 9
177176

Hope this helps

Dave

**Reply by alex-code** [reply]
For the menu icon to register a touch `x` needs to be between 290 & 320 and `y` between 0 & 30.

If you keep touching around the top right can you get `x` above 290?

**Reply by daveclifford** [reply]
NO
**On Screen Rotation**  screen
touching upper right trying to get to top of screen touch area this is best I can do - x=282


X: 282 - Y: 10
177272
X: 282 - Y: 10
177272
X: 282 - Y: 4
177272
X: 282 - Y: 1
177272
X: 282 - Y: 12
177272
X: 281 - Y: 10
X: 281 - Y: 10
X: 282 - Y: 1
177176
X: 282 - Y: 19
177176
X: 282 - Y: 19
177176
X: 282 - Y: 20
177176
X: 282 - Y: 9
177176
X: 282 - Y: 2
177176
X: 282 - Y: 1
177176
X: 282 - Y: 11
177176
X: 282 - Y: 11
177176
X: 282 - Y: 11
177176
X: 282 - Y: 20
177176
X: 282 - Y: 20
177176
X: 282 - Y: 20
177176
X: 281 - Y: 29
X: 282 - Y: 25
177176
X: 282 - Y: 25
177176
X: 282 - Y: 30
177176
X: 282 - Y: 30
177176
X: 282 - Y: 32
177176
X: 282 - Y: 32
177176
X: 282 - Y: 7
177176
X: 282 - Y: 7
177176
X: 282 - Y: 3
177176
X: 282 - Y: 4
177176
X: 282 - Y: 4
177176
X: 282 - Y: 3
177176
X: 282 - Y: 5
177176
X: 282 - Y: 2
177176
X: 282 - Y: 4
177176
X: 282 - Y: 10
177176
X: 282 - Y: 3
177176
X: 282 - Y: 1
177176
X: 282 - Y: 1
177176
X: 282 - Y: 1
177176
X: 282 - Y: 1
177176
X: 282 - Y: 1
177176
X: 282 - Y: 1
177176
X: 282 - Y: 1
177176
X: 280 - Y: 1
177176
X: 281 - Y: 1
177176
X: 281 - Y: 1
177176
X: 282 - Y: 1
177176

Are my screen dimensions a little different so that my values are outside your parameter for the target area?

Should I disassemble the throttle to reseat the ribbon cable? 

I don't think the cable  is at fault, since the screen is displaying correctly and touching the buttons  for the most part, works. 

Thanks much for your guidance and assistance.

Dave

**Reply by daveclifford** [reply]
I just noticed there  is a  black  band (screen not lit looking lighter) at the top of the touch screen.

The touchscreen dimensions are 55mm wide x 85mm high.

There is a black band 8mm high across the screnn that does not appear active (black - not lit  looking lighter as the active screen shows).

Is this correct? 

Does your throttle screen have a black band at the top?

Dave

**Reply by daveclifford** [reply]
Module Size - W x H x T	95 mm x 55.5 mm  (I only see 85 mm)
Viewing Area - W x H	73.44 mm x 48.96 mm

The touchscreen  viewing area (according to the specs) is 73.44 mm x 48.96 mm

85mm - 8mm (the black band at top)  = 77mm (close enough to 73.44 mm)

I'm going wacky trying to figure out this issue

Dave

**Reply by alex-code** [reply]
Those dimensions look ok.
There is a black border around the display, it doesn't fill the whole glass bit.

One other thing to check is the resolution the touch screen has.
After line https://github.com/DCC-EX/EX-T3-WiFi/blob/master/src/main.cpp#L316 you could add the below so we can check the x resolution.
`Serial.println(cfg->xResolution);`

**Reply by daveclifford** [reply]
Wait for it, wait for it .......   the number is ...........  320  (is that the number  you were thinking of?)

 I have an accelerometer attached and noticed that  it doesn't flip the screen.

Dave

**Reply by alex-code** [reply]
320 is what should be expected.

The problem is if you can't exit from settings back to the menu without power cycling, the settings (like screen rotation) won't be saved.

Leaving in the Serial outputs for `x` & `y` what `x` do you get if you touch anywhere on the right side of the display.

**Reply by daveclifford** [reply]
X: 282 - Y: 408
156548

**Reply by daveclifford** [reply]
weird - original in spanish

**Reply by alex-code** [reply]
I'm lost on this one, no idea why `x` won't go above 282.
Might have to make a debug app, something I've seen before where it draws as you touch the screen so we can see if the coords relate to the  x & y of the display.

**Comment by daveclifford** [comment]
I plan to take the throttle partially apart to get to the touchscreen cable - reseat it - and then reassemble

I also think the accelerometer connections (4 solid wires ) may not be making contact with th FireBeetle2.
I can't see the screen flip when turning the throttle clockwise /counterclockwise.

For me,  it's worth a try in disassembling my pride and joy.

A quick question -  I assume these debug lines only reduce memory for real code, but won't affect the operation, since the user only sees the result in monitor mode, If so,  why don't you put some instructions into the code for this purpose? You can take them out if program memory becomes tight.

I  will do the disassembly/reassembly tomorrow Wednesday 4/5.

Dave

**Reply by alex-code** [reply]
A few Serial prints won't affect much.
They're only really needed for this kind of debugging but I've not encountered this issue so not needed them.

When you're on the settings could you invert the display then touch the same top right corner and see what x & y that shows.

**Comment by daveclifford** [comment]
inverted 
push lower left
X: 302 - Y: 0
X: 289 - Y: 21
X: 294 - Y: 12
X: 296 - Y: 6

now push upper right *still inverted)
X: 43 - Y: 447
155948

X: 300 - Y: 9
push lower left

X: 42 - Y: 449
155956

Dave
